### ISSUE TITLE
unnecessary_type_check false positive

### ISSUE THREAD

**Issue by nt4f04uNd** [dart-lang/linter#3133]
```dart
void main() {
  final B<A> someObject = B();
  assert(someObject is B<A> || someObject is B<A1>);
}

class A {}
class A1 implements A {}
class B<T extends A> {}
```

Linter suggests that `someObject is B<A>` is unnecessary, however the assertion would fail without this check.

**Comment by nt4f04uNd** [comment]
I just discovered that `someObject is B<A>` is now always true in null-safe world, previously it wasn't.
